syft/frameworks/torch/nn/rnn.py

```diff
@@ -0,0 +1,399 @@
+                            h_next[layer, :, :], c_next[layer, :, :] = self.rnn_backward[layer](
+                                h_next[layer - 1, :, :],
+                                (h_backward[layer, :, :], c_backward[layer, :, :]),
+                            )
+                        else:
+                            h_next[layer, :, :] = self.rnn_backward[layer](
+                                h_next[layer - 1, :, :], h_backward[layer, :, :]
+                            )
+                output_backward[t, :, :] = h_next[layer, :, :]
+                h_backward = h_next
+                if self.is_lstm:
+                    c_backward = c_next
+
+            output = torch.cat((output, output_backward), dim=-1)
+            hidden = torch.cat((hidden, h_backward), dim=0)
+            if self.is_lstm:
+                cell = torch.cat((cell, c_backward), dim=0)
+
+        if self.batch_first:
+            output = torch.transpose(output, 0, 1)
+            hidden = torch.transpose(hidden, 0, 1)
+            if self.is_lstm:
+                cell = torch.transpose(cell, 0, 1)
+
+        if self.is_lstm:
+            hidden = (hidden, cell)
+
+        return output, hidden
+
+
+class RNN(RNNBase):
+    """
+    Python implementation of RNN for MPC
+    This class overrides the torch.nn.RNN
+    """
+
+    def __init__(
+        self,
+        input_size,
+        hidden_size,
+        num_layers=1,
+        nonlinearity="tanh",
+        bias=True,
+        batch_first=False,
+        dropout=0,
+        bidirectional=False,
+    ):
+
+        super(RNN, self).__init__(
+            input_size,
+            hidden_size,
+            num_layers,
+            bias,
+            batch_first,
+            dropout,
+            bidirectional,
+            RNNCell,
+            nonlinearity,
+        )
+
+
+class GRU(RNNBase):
+    """
+    Python implementation of GRU for MPC
+    This class overrides the torch.nn.GRU
+    """
+
+    def __init__(
+        self,
+        input_size,
+        hidden_size,
+        num_layers=1,
+        bias=True,
+        batch_first=False,
+        dropout=0,
+        bidirectional=False,
+    ):
+
+        super(GRU, self).__init__(
+            input_size, hidden_size, num_layers, bias, batch_first, dropout, bidirectional, GRUCell
+        )
+
+
+class LSTM(RNNBase):
+    """
+    Python implementation of LSTM for MPC
+    This class overrides the torch.nn.LSTM
+    """
+
+    def __init__(
+        self,
+        input_size,
+        hidden_size,
+        num_layers=1,
+        bias=True,
+        batch_first=False,
+        dropout=0,
+        bidirectional=False,
+    ):
+
+        super(LSTM, self).__init__(
+            input_size, hidden_size, num_layers, bias, batch_first, dropout, bidirectional, LSTMCell
+        )
+
+
+def init_hidden(input, hidden_size, n_layers=None):
+    if n_layers is None:
+        h = torch.zeros(input.shape[0], hidden_size, dtype=input.dtype, device=input.device)
```

Comment:
Just checking... shouldn't this be input.shape[1] instead of input.shape[0]?

Reply:
here input.shape[0] is the batch_size and input.shape[1] is the embedding size (exactly as in the PyTorch original RNNCell module: https://pytorch.org/docs/stable/nn.html#torch.nn.RNNCell)

We need to initialize h with shape (batch_size, hidden_size)